### PR TITLE
Handle NULL string results for some metadata queries

### PR DESCRIPTION
**Pull request by jimmyyih**
After a couple GPDB patches were merged into GPDB 6X and 5X, we no
longer get cache errors for a handful of metadata queries. Instead, we
get result parsing errors because we store those results into strings
that do not accept NULL values. To handle NULL values, we replace all
related string result variables into sql.NullString variables. A
validation check is also added to the end of each query function to
remove all objects that contain these NULL values to ensure the
metadata backup is pristine. Added some unit tests as well to validate
the validation check and updated a lot of tests to handle the string
to sql.NullString change.

GPDB Commit References:
https://github.com/greenplum-db/gpdb/commit/1ccc3df7630646fa409b65a3c9a6e15c8ee2e1a0
https://github.com/greenplum-db/gpdb/commit/00377e023737ccbe98368396be651be1d72922e1

### CODE REVIEW

File: backup/queries_functions.go [hunk 1]
```diff
@@ -143,7 +147,20 @@ func GetFunctions(connectionPool *dbconn.DBConn) []Function {
...
+	// Remove all functions that have NULL arguments, NULL identity
+	// arguments, or NULL result type. This can happen if the query
+	// above is run and a concurrent function drop happens just
+	// before the pg_get_function_* functions execute.
+	verifiedResults := make([]Function, 0)
+	for _, result := range results {
+		if result.Arguments.Valid && result.IdentArgs.Valid && result.ResultType.Valid {
+			verifiedResults = append(verifiedResults, result)
+		} else {
+			gplog.Warn("Invalid backup for function '%s.%s'. It was most likely concurrently dropped.", result.Schema, result.Name)
```
Comment (jimmyyih) [inline backup/queries_functions.go:159]: @oakbarrett Please review this new warning message.

File: backup/queries_functions.go [hunk 2]
```diff
@@ -484,29 +501,47 @@ func GetAggregates(connectionPool *dbconn.DBConn) []Aggregate {
...
+		// Remove all aggregates that have NULL arguments or NULL
+		// identity arguments. This can happen if the query above
+		// is run and a concurrent aggregate drop happens before
+		// the pg_get_function_* functions execute.
+		verifiedAggregates := make([]Aggregate, 0)
+		for _, aggregate := range aggregates {
+			if aggregate.Arguments.Valid && aggregate.IdentArgs.Valid {
+				verifiedAggregates = append(verifiedAggregates, aggregate)
+			} else {
+				gplog.Warn("Invalid backup for aggregate '%s.%s'. It was most likely concurrently dropped.", aggregate.Schema, aggregate.Name)
```
Comment (jimmyyih) [inline backup/queries_functions.go:521]: @oakbarrett Please review this new warning message.

File: backup/queries_functions.go [hunk 3]
```diff
@@ -541,14 +576,21 @@ func GetFunctionOidToInfoMap(connectionPool *dbconn.DBConn) map[uint32]FunctionI
...
+			results[i].IdentArgs.String = arguments[results[i].Oid]
+			results[i].IdentArgs.Valid = true // Hardcode for GPDB 4.3 to fit sql.NullString
 		}
 	} else {
 		err = connectionPool.Select(&results, query)
 	}
 	gplog.FatalOnError(err)
 	for _, funcInfo := range results {
+		if !funcInfo.Arguments.Valid || !funcInfo.IdentArgs.Valid {
+			gplog.Warn("Invalid backup for function '%s.%s'. It was most likely concurrently dropped.", funcInfo.Schema, funcInfo.Name)
```
Comment (jimmyyih) [inline backup/queries_functions.go:590]: @oakbarrett Please review this new warning message.

File: backup/queries_postdata.go [hunk 4]
```diff
@@ -143,15 +144,29 @@ func GetIndexes(connectionPool *dbconn.DBConn) []IndexDefinition {
...
+	// Remove all indexes that have NULL definitions. This can happen
+	// if a concurrent index drop happens before the associated table
+	// lock is acquired earlier during gpbackup execution.
+	verifiedResultIndexes := make([]IndexDefinition, 0)
+	for _, resultIndex := range resultIndexes {
+		if resultIndex.Def.Valid {
+			verifiedResultIndexes = append(verifiedResultIndexes, resultIndex)
+		} else {
+			gplog.Warn("Invalid backup for index '%s' on table '%s.%s'. It was most likely concurrently dropped.",
+				resultIndex.Name, resultIndex.OwningSchema, resultIndex.OwningTable)
```
Comment (jimmyyih) [inline backup/queries_postdata.go:157]: @oakbarrett Please review this new warning message.

File: backup/queries_postdata.go [hunk 5]
```diff
@@ -200,7 +215,21 @@ func GetRules(connectionPool *dbconn.DBConn) []RuleDefinition {
...
+	// Remove all rules that have NULL definitions. Not sure how
+	// this can happen since pg_get_ruledef uses an SPI query but
+	// handle the NULL just in case.
+	verifiedResults := make([]RuleDefinition, 0)
+	for _, result := range results {
+		if result.Def.Valid {
+			verifiedResults = append(verifiedResults, result)
+		} else {
+			gplog.Warn("Invalid backup for rule '%s' on table '%s.%s'. It was most likely concurrently dropped.",
+				result.Name, result.OwningSchema, result.OwningTable)
```
Comment (jimmyyih) [inline backup/queries_postdata.go:228]: @oakbarrett Please review this new warning message.

File: backup/queries_postdata.go [hunk 6]
```diff
@@ -250,7 +279,21 @@ func GetTriggers(connectionPool *dbconn.DBConn) []TriggerDefinition {
...
+	// Remove all triggers that have NULL definitions. This can happen
+	// if the query above is run and a concurrent trigger drop happens
+	// just before the pg_get_triggerdef function executes.
+	verifiedResults := make([]TriggerDefinition, 0)
+	for _, result := range results {
+		if result.Def.Valid {
+			verifiedResults = append(verifiedResults, result)
+		} else {
+			gplog.Warn("Invalid backup for trigger '%s' on table '%s.%s'. It was most likely concurrently dropped.",
+				result.Name, result.OwningSchema, result.OwningTable)
```
Comment (jimmyyih) [inline backup/queries_postdata.go:292]: @oakbarrett Please review this new warning message.

File: backup/queries_relations.go [hunk 7]
```diff
@@ -334,7 +335,19 @@ func GetAllViews(connectionPool *dbconn.DBConn) []View {
...
-	return results
+	// Remove all views that have NULL definitions. This can happen
+	// if the query above is run and a concurrent view drop happens
+	// just before the pg_get_viewdef function execute.
+	verifiedResults := make([]View, 0)
+	for _, result := range results {
+		if result.Definition.Valid {
+			verifiedResults = append(verifiedResults, result)
+		} else {
+			gplog.Warn("Invalid backup for view '%s.%s'. It was most likely concurrently dropped.", result.Schema, result.Name)
```
Comment (jimmyyih) [inline backup/queries_relations.go:346]: @oakbarrett Please review this new warning message.

File: backup/queries_shared.go [hunk 8]
```diff
@@ -174,7 +175,26 @@ func GetConstraints(connectionPool *dbconn.DBConn, includeTables ...Relation) []
...
+		// just before the pg_get_constraintdef function executes. Note that
+		// GPDB 6+ pg_get_constraintdef uses an MVCC snapshot instead of
+		// syscache so we only need to verify for GPDB 4.3 and 5 (technically
+		// only GPDB 5 since 4.3 will get a cache error on the query).
+		verifiedResults := make([]Constraint, 0)
+		for _, result := range results {
+			if result.ConDef.Valid {
+				verifiedResults = append(verifiedResults, result)
+			} else {
+				gplog.Warn("Invalid backup for constraint '%s.%s'. It was most likely concurrently dropped.", result.Schema, result.Name)
```
Comment (jimmyyih) [inline backup/queries_shared.go:191]: @oakbarrett Please review this new warning message.

Comment (oakbarrett) [reply backup/queries_shared.go:191]: Constraint '%s.%s' not backed up, most likely dropped after gpbackup had begun.

File: backup/queries_relations.go (hunk 7 above)
Comment (oakbarrett) [reply backup/queries_relations.go:346]: View '%s.%s' not backed up, most likely dropped after gpbackup had begun.

File: backup/queries_postdata.go (hunk 6 above)
Comment (oakbarrett) [reply backup/queries_postdata.go:292]: Trigger '%s' on table '%s.%s' not backed up, most likely dropped after gpbackup had begun.

File: backup/queries_postdata.go (hunk 5 above)
Comment (oakbarrett) [reply backup/queries_postdata.go:228]: Rule '%s' on table '%s.%s' not backed up, most likely dropped after gpbackup had begun.

File: backup/queries_postdata.go (hunk 4 above)
Comment (oakbarrett) [reply backup/queries_postdata.go:157]: Index '%s.%s' on table '%s.%s' not backed up, most likely dropped after gpbackup had begun.

File: backup/queries_functions.go (hunk 3 above)
Comment (oakbarrett) [reply backup/queries_functions.go:590]: Function '%s.%s' not backed up, most likely dropped after gpbackup had begun.

File: backup/queries_functions.go (hunk 2 above)
Comment (oakbarrett) [reply backup/queries_functions.go:521]: Aggregate '%s.%s' not backed up, most likely dropped after gpbackup had begun.

File: backup/queries_functions.go (hunk 1 above)
Comment (oakbarrett) [reply backup/queries_functions.go:159]: @jimmyyih 

Function '%s.%s' not backed up, most likely dropped after gpbackup had begun.

File: backup/queries_relation_internal_test.go [hunk 9]
```diff
@@ -1,33 +1,55 @@
-package backup
+package backup_test
```
Comment (kyeap-pivotal) [inline backup/queries_relation_internal_test.go:1]: I've been following a convention of naming files that include tests in the package with *_internal_tests.go. Now that we are going to put tests outside the package we can rename this file to queries_relation_test.go.

Comment (jimmyyih) [reply backup/queries_relation_internal_test.go]: Renamed the file.

**Pull request merged**